File: chaos-flutter/lib/core/backend/chaos_backend.dart
import '../models/danmaku.dart';
import '../models/live.dart';
import '../models/live_directory.dart';
import '../models/lyrics.dart';
import '../models/music.dart';
import '../models/now_playing.dart';
import '../models/subtitles.dart';

abstract class ChaosBackend {
  String get name;

  // Live directory.
  Future<List<LiveDirCategory>> categories(String site);
  Future<LiveDirRoomListResult> recommendRooms(String site, int page);
  Future<LiveDirRoomListResult> categoryRooms(
    String site,
    String? parentId,
    String categoryId,
    int page,
  );
  Future<LiveDirRoomListResult> searchRooms(
      String site, String keyword, int page);

  // Live playback.
  Future<LivestreamDecodeManifestResult> decodeManifest(String input);
  Future<LiveOpenResult> openLive(String input, {String? variantId});
  Future<void> closeLive(String sessionId);

  Stream<DanmakuMessage> danmakuStream(String sessionId);
  Future<DanmakuFetchImageResult> fetchDanmakuImage(
      String sessionId, String url);

  // Lyrics / now playing.
  Future<NowPlayingSnapshot> nowPlayingSnapshot(
      NowPlayingSnapshotParams params);
  Future<List<LyricsSearchResult>> lyricsSearch(LyricsSearchParams params);

  // Music.
  Future<void> musicConfigSet(MusicProviderConfig cfg);
  Future<List<MusicTrack>> searchTracks(MusicSearchParams p);
  Future<List<MusicAlbum>> searchAlbums(MusicSearchParams p);
  Future<List<MusicArtist>> searchArtists(MusicSearchParams p);
  Future<List<MusicTrack>> albumTracks(MusicAlbumTracksParams p);
  Future<List<MusicAlbum>> artistAlbums(MusicArtistAlbumsParams p);
  Future<MusicTrackPlayUrlResult> trackPlayUrl(MusicTrackPlayUrlParams p);
  Future<MusicLoginQr> qqLoginQrCreate(String loginType);
  Future<MusicLoginQrPollResult> qqLoginQrPoll(String sessionId);
  Future<QqMusicCookie> qqRefreshCookie(QqMusicCookie cookie);
  Future<MusicLoginQr> kugouLoginQrCreate(String loginType);
  Future<MusicLoginQrPollResult> kugouLoginQrPoll(String sessionId);
  Future<MusicDownloadStartResult> downloadStart(MusicDownloadStartParams p);
  Future<MusicDownloadStatus> downloadStatus(String sessionId);
  Future<void> downloadCancel(String sessionId);

  // Subtitles.
  Future<List<ThunderSubtitleItem>> subtitleSearch(SubtitleSearchParams p);
  Future<SubtitleDownloadReply> subtitleDownload(SubtitleDownloadParams p);

  Future<void> dispose();
}

File: chaos-android/app/src/main/java/com/zerodevi1/chaos_seed/core/backend/ChaosBackend.kt
package com.zerodevi1.chaos_seed.core.backend

import com.zerodevi1.chaos_seed.core.model.DanmakuMessage
import com.zerodevi1.chaos_seed.core.model.LiveDirCategory
import com.zerodevi1.chaos_seed.core.model.LiveDirRoomListResult
import com.zerodevi1.chaos_seed.core.model.LivestreamDecodeManifestResult
import com.zerodevi1.chaos_seed.core.model.LiveOpenResult
import com.zerodevi1.chaos_seed.core.model.LivestreamVariant
import com.zerodevi1.chaos_seed.core.model.LyricsSearchParams
import com.zerodevi1.chaos_seed.core.model.LyricsSearchResult
import com.zerodevi1.chaos_seed.core.model.MusicDownloadStartParams
import com.zerodevi1.chaos_seed.core.model.MusicDownloadStartResult
import com.zerodevi1.chaos_seed.core.model.MusicDownloadStatus
import com.zerodevi1.chaos_seed.core.model.MusicLoginQr
import com.zerodevi1.chaos_seed.core.model.MusicLoginQrPollResult
import com.zerodevi1.chaos_seed.core.model.MusicProviderConfig
import com.zerodevi1.chaos_seed.core.model.MusicSearchParams
import com.zerodevi1.chaos_seed.core.model.MusicTrack
import kotlinx.coroutines.flow.Flow

interface ChaosBackend {
    val name: String

    // Live directory.
    suspend fun categories(site: String): List<LiveDirCategory>
    suspend fun recommendRooms(site: String, page: Int): LiveDirRoomListResult
    suspend fun categoryRooms(site: String, parentId: String?, categoryId: String, page: Int): LiveDirRoomListResult
    suspend fun searchRooms(site: String, keyword: String, page: Int): LiveDirRoomListResult

    // Live playback.
    suspend fun decodeManifest(input: String): LivestreamDecodeManifestResult
    suspend fun resolveVariant2(site: String, roomId: String, variantId: String): LivestreamVariant
    suspend fun openLive(input: String, variantId: String? = null): LiveOpenResult
    suspend fun closeLive(sessionId: String)
    fun danmakuStream(sessionId: String): Flow<DanmakuMessage>

    // Music.
    suspend fun musicConfigSet(cfg: MusicProviderConfig)
    suspend fun searchTracks(p: MusicSearchParams): List<MusicTrack>
    suspend fun qqLoginQrCreate(loginType: String): MusicLoginQr
    suspend fun qqLoginQrPoll(sessionId: String): MusicLoginQrPollResult
    suspend fun downloadStart(p: MusicDownloadStartParams): MusicDownloadStartResult
    suspend fun downloadStatus(sessionId: String): MusicDownloadStatus
    suspend fun downloadCancel(sessionId: String)

    // Lyrics.
    suspend fun lyricsSearch(p: LyricsSearchParams): List<LyricsSearchResult>
}

